### ISSUE TITLE
cache the AST of each GR(1) clause

### ISSUE THREAD

**Issue by johnyf** [tulip-control/tulip-control#100]
Parsing of huge clauses can consume some time. With `pyparsing` that time was unacceptable, but with `PLY` it is typically unnoticed. In any case, as of 7048530f0884ce3d1e028b27bae9a98f9fc11a68 calls `to_gr1c` and analogous functions can occur multiple times within `GRSpec`. So storing the resulting AST seems like a good idea to reduce duplication of computation.

This can be done for example by maintaining a `dict` whose keys are `GRSpec` clauses (elements in the containers `env_init`, `env_safety`, etc.) and values the corresponding ASTs. If a `GRSpec` is changed, then whichever clauses changed won't be found in the `dict` and will be parsed at that time and added to the `dict`.

Such absence of clauses from the `dict` will also trigger a removal of clause-AST pairs that are not included in the `GRSpec` any more. It is expected that typically parsing will occur only once (we rarely remove clauses from formulas).

**Comment by slivingston** [comment]
I want to propose an alternative term to describe the feature introduced in this issue: "memoization".  "Cache" seems too general, motivating my question [above](https://github.com/tulip-control/tulip-control/issues/100#issuecomment-53502364) in reply to the opening post. Usually a function `f` is said to be memoizing when it saves a (hidden) copy of results associated with a particular setting of its parameters so that, given the same parameter values, it can return the saved result and avoid repeated computation.

**Comment by slivingston** [comment]
I was able to repeat the timing experiment using the script [above](https://github.com/tulip-control/tulip-control/issues/100#issuecomment-53672293) against commits 265f54ec7486c99e58a43d6986e1ddae04347600 and 0ae146b2b02149ff27e3e7f59546d6c91de91a3e , and I observed similar ratios of times, thus supporting the above conclusions about performance improvement.

For posterity, the timing script given [above](https://github.com/tulip-control/tulip-control/issues/100#issuecomment-53672293) is updated for the current `tulip.spec` as follows:

```python
from tulip import spec

env_vars = {'a':'boolean'}
env_prog = ['a -> b']

sys_vars = {'b':(0,5), 'c':['hehe', 'ho', 'haha']}
sys_prog = ['(b = 0) && (c = "ho") && []<>(b > 0 || X (c = "hehe") )']

spc = spec.GRSpec(env_vars=env_vars, sys_vars=sys_vars,
                  env_prog=env_prog, sys_prog=sys_prog)
N = 5
for i in xrange(10**N):
    s = spec.translation.translate(spc, 'gr1c')
```

where the formula is also improved to make it well-formed (e.g., in the original script, what does `next(c)` mean?). Again, the improvement in performance is observed.
